mesonbuild/interpreter/interpreter.py

```diff
@@ -1681,11 +1667,40 @@ def program_lookup(self, args: T.List[mesonlib.FileOrString], for_machine: Machi
         if progobj is None and args[0].endswith('python3'):
             prog = ExternalProgram('python3', mesonlib.python_command, silent=True)
             progobj = prog if prog.found() else None
+
+        if progobj and not self.check_program_version(progobj, wanted, version_func, extra_info):
+            progobj = None
+
         if progobj is None and fallback and required:
```

Comment:
Is checking `and required` here correct? Wouldn't I want to fallback if regardless of whether required is True or not?

Reply:
It's a bit surprising but that's intentional, dependency() does that same. This allows doing manual fallbacks to other system programs before using a subproject:
```meson
prog = find_program('foo-2.0', required: false)
if not prog.found()
  prog = find_program('foo-1.0', required: false)
endif
if not prog.found()
  # This will do the fallback
  prog = find_program('foo-2.0')
endif
```

Probably not really useful for programs as that can be done as `find_program('foo-2.0', 'foo-1.0')` but I think it's better to keep consistent with dependency().

Reply:
That was already like that before this PR, so if we want to change that behaviour that should be a different PR anyway.